core/src/main/java/org/elasticsearch/common/settings/Settings.java

```diff
@@ -1032,7 +1032,110 @@ public Builder normalizePrefix(String prefix) {
+
+                @Override
+                public Iterator<Entry<String, String>> iterator() {
+                    Iterator<Entry<String, String>> iter = delegateSet.iterator();
+
+                    return new Iterator<Entry<String, String>>() {
+                        private int numIterated;
+                        Entry<String, String> current;
+                        @Override
+                        public boolean hasNext() {
+                            if (numIterated == size) { // early terminate
+                                assert size != -1 : "size was never set: " + numIterated + " vs. " + size;
+                                return false;
+                            }
+                            while(iter.hasNext()) {
+                                if (filter.test((current = iter.next()).getKey())) {
+                                    numIterated++;
+                                    return true;
+                                }
+                            }
+                            current = null;
+                            return false;
+                        }
+
+                        @Override
+                        public Entry<String, String> next() {
+                            if (prefix == null) {
+                                return current;
+                            }
+                            return new Entry<String, String>() {
+                                @Override
+                                public String getKey() {
+                                    return current.getKey().substring(prefix.length());
+                                }
+
+                                @Override
+                                public String getValue() {
+                                    return current.getValue();
+                                }
+
+                                @Override
+                                public String setValue(String value) {
+                                    throw new UnsupportedOperationException();
+                                }
+                            };
+                        }
+                    };
+                }
+
+                @Override
+                public int size() {
+                    if (size == -1) {
+                        size = Math.toIntExact(delegateSet.stream().filter((e) -> filter.test(e.getKey())).count());
+                    }
+                    return size;
+                }
+            };
+            return filterSet;
+        }
+
+        private FilteredMap(Map<String, String> delegate, Predicate<String> filter, String prefix) {
+            this.delegate = delegate;
+            this.filter = filter;
+            this.prefix = prefix;
+        }
+
+        @Override
+        public String get(Object key) {
+            String theKey = prefix == null ? key.toString() : prefix + key;
```

Comment:
I think we should return `null` if the key is not an instance of `String` rather than converting it